docs/reconfigurator.adoc

```diff
@@ -179,6 +179,166 @@ We're being cautious about rolling out that kind of automation.  Instead, today,
+* In the medium term, for online upgrade, we will be incorporating an image id or artifact id into the Omicron zone configuration.  Currently, the artifact id is implied: sled agent uses whatever artifact was delivered by the last MUPdate.  For online upgrade, the control plane will need to be able to specify a particular artifact.
+
+In all of these cases:
+
+* There's a piece of configuration managed outside of CockroachDB/Nexus (e.g., by Sled Agent, RSS, and/or MUPdate).
+* We want to transition to a world where the configuration is owned by CockroachDB/Nexus.
+* We need to bootstrap the initial control-plane-managed copy based on what's currently deployed.
+
+The general pattern here is:
+
+* Use the inventory system to collect the current configuration.
+* Incorporate that configuration into the next blueprint generated by Reconfigurator.
+* Incorporate the configuration from the blueprint into the next request to Sled Agent.
+
+```mermaid
+sequenceDiagram
+    participant SledAgent as Sled Agent
+    participant Nexus
+
+    Note over SledAgent: owns some piece<br/> of configuration
+    SledAgent ->> Nexus: reports current configuration<br/>via inventory
+    Note over Nexus: incorporates latest inventory<br/>into next blueprint
+    Note over Nexus: now owns the configuration
+
+    loop
+	Note over Nexus: changes configuration as needed
+        Nexus ->> SledAgent: sends new configuration<br/>
+    end
+```
+
+Below is a proposed pattern for doing this over two releases.  We'll call the new piece of config `my_config`, represented with type `MyConfig`.  This could be arbitrarily complicated.  In our examples above, this could be a list of zones and their detailed configurations (IP addresses, ports, and all the properties they need to start), a list of ZFS dataset structs, a `dataset_id` property on an existing struct, an `artifact_id` property on an existing struct, etc.  It may hang directly off the Sled Agent `Inventory` or it might be embedded in some existing struct.
+
+NOTE: This is a work in progress.  We hope this closely enough matches what we've done in the past that it should work.  We should update this documentation as we discover better ways to do this.
+
+CAUTION: This isn't the _only_ way to do it.  However, many other ways to do it come with non-obvious problems.  When we diverge from this, we should first try to understand why this procedure looks the way it does.
+
+**In the first release** (we'll call it "release 1"): the configuration is totally managed in Sled Agent and unknown to Nexus and CockroachDB.
+
+**In the next release** (we'll call it "release 2"):
+
+. Add `my_config: MyConfig` to appropriate spot in Sled Agent inventory.  Note that it is _not_ an optional field.  In this use case, Sled Agent always knows the current value of this field.
```

Comment:
So, in a case like adding the artifact/image ID, we're also adding the requirement that sled agent needs to learn about this value explicitly, right?

As an example -- "release 1" of the sled agent might launch instances using `/opt/oxide/propolis-server.tar.gz` -- and use whatever file that is, regardless of "version".

Just wanted to call that out, as "part of release 2 may involve making sled agent explicitly aware of what state it happens to be managing", since we've declared it should be non-optional in inventory.

(also, to be clear, it's non-optional in the **sled agent API call to Nexus**, not the inventory database structure, right? Otherwise, old inventory collections would be invalidated)

Reply:
> So, in a case like adding the artifact/image ID, we're also adding the requirement that sled agent needs to learn about this value explicitly, right?
> 
> As an example -- "release 1" of the sled agent might launch instances using `/opt/oxide/propolis-server.tar.gz` -- and use whatever file that is, regardless of "version".
> 
> Just wanted to call that out, as "part of release 2 may involve making sled agent explicitly aware of what state it happens to be managing", since we've declared it should be non-optional in inventory.


The code needs to be aware that this value, which may previously have been hardcoded or even absent altogether, is now a variable to be reported in inventory (this item) and controlled by Nexus (item 4).  Is that what you mean by "learn about this value explicitly"?

> (also, to be clear, it's non-optional in the **sled agent API call to Nexus**, not the inventory database structure, right? Otherwise, old inventory collections would be invalidated)

Good point!  I will update this to be more clear that (1) the inventory API structure can make this required but the inventory database structures should have this value optional, and (2) in release 3, this can probably be made required in the inventory database structures, too.  (This is annoyingly tricky: we _might_ still have old collections at this point but it seems unlikely.)

Reply:
> The code needs to be aware that this value, which may previously have been hardcoded or even absent altogether, is now a variable to be reported in inventory (this item) and controlled by Nexus (item 4). Is that what you mean by "learn about this value explicitly"?

Yeah, basically just calling this out -- the phrasing "In this use case, Sled Agent always knows the current value of this field" is true, but if it was an "implicit" value before, Sled Agent might need to do some legwork to get an explicit value that can be reported.

> <Inventory stuff>

Yeah, totally - another spot where it might be useful to somehow provide a barrier. If we could ensure that all inventory collections are "from this release or the one prior, and no older", that could be a useful way to gradually make these values non-optional.

Reply:
> > (This is annoyingly tricky: we might still have old collections at this point but it seems unlikely.)
> 
> Is this almost identical (or even exactly identical) to "changing an optional field to required in a blueprint requires dropping old blueprints", with the exception that inventory already rotates old collections out with relatively high frequency except in the case of errors (where we keep around the last good collection, which might be months old on a system like dogfood)? Do you think we #7278 should explicitly include inventory collections too? (It already implicitly covers them to some extent, since saving the reconfigurator state already includes them, but doesn't mention deleting old ones if they're still around.)

Yes, it's analogous.  Sure -- I'll update that issue.